src/librustc_middle/ty/print/pretty.rs

```diff
@@ -882,29 +894,39 @@ pub trait PrettyPrinter<'tcx>:
 
         match ct.val {
             ty::ConstKind::Unevaluated(did, substs, promoted) => {
-                if let Some(promoted) = promoted {
-                    p!(print_value_path(did, substs));
-                    p!(write("::{:?}", promoted));
-                } else {
-                    match self.tcx().def_kind(did) {
-                        DefKind::Static | DefKind::Const | DefKind::AssocConst => {
-                            p!(print_value_path(did, substs))
-                        }
-                        _ => {
-                            if did.is_local() {
-                                let span = self.tcx().def_span(did);
-                                if let Ok(snip) = self.tcx().sess.source_map().span_to_snippet(span)
-                                {
-                                    p!(write("{}", snip))
+                // Don't re-entrantly enter this code path: that is,
+                // don't try to print the DefPath of an unevaluated const
+                // while we're already printing the DefPath of an unevaluated
+                // const. This ensures that we never end up trying to recursively
+                // print a const that we're already printing.
+                // See issue #68104 for more details
+                self = with_no_const_path(|| {
```

Comment:
Can you try `with_forced_impl_filename_line` instead? It should also work since I believe the problem here is the pretty-printing of `impl` paths.

Wait. Hang on. Why would this *ever* print paths?!

The bug is probably in metadata encoding/decoding. `AnonConst`s shouldn't turn into `DefKind::Const`.